src/fonduer/supervision/labeler.py

```diff
@@ -0,0 +1,237 @@
+import logging
+
+from fonduer.candidates.models import Candidate
+from fonduer.supervision.models import GoldLabelKey, Label, LabelKey
+from fonduer.utils.udf import UDF, UDFRunner
+from fonduer.utils.utils_udf import (
+    ALL_SPLITS,
+    add_keys,
+    batch_upsert_records,
+    get_cands_list_from_split,
+    get_docs_from_split,
+    get_mapping,
+    get_sparse_matrix,
+)
+
+logger = logging.getLogger(__name__)
+
+
+def load_gold_labels(session, cand_lists, annotator_name="gold"):
+    """Load the sparse matrix for the specified annotator."""
+    return get_sparse_matrix(session, GoldLabelKey, cand_lists, key=annotator_name)
+
+
+class Labeler(UDFRunner):
+    """An operator to add Label Annotations to Candidates."""
+
+    def __init__(self, session, candidate_classes):
+        """Initialize the Labeler."""
+        super(Labeler, self).__init__(
+            session, LabelerUDF, candidate_classes=candidate_classes
+        )
+        self.candidate_classes = candidate_classes
+        self.lfs = []
+
+    def update(self, docs=None, split=0, lfs=None, **kwargs):
+        """Call apply with update=True."""
+        if lfs is None:
+            raise ValueError("Please provide a list of lists of labeling functions.")
+
+        if len(lfs) != len(self.candidate_classes):
+            raise ValueError("Please provide LFs for each candidate class.")
+
+        for i in range(len(self.lfs)):
+            # Filter out the new/updated LFs
+            self.lfs[i] = [
+                lf
+                for lf in self.lfs[i]
+                if lf.__name__ not in [_.__name__ for _ in lfs[i]]
+            ]
+            # Then add them
+            self.lfs[i].extend(lfs[i])
+
+        self.apply(
+            docs=docs,
+            split=split,
+            lfs=self.lfs,
+            train=False,
+            update=True,
+            clear=False,
+            **kwargs
+        )
+
+    def apply(
+        self,
+        docs=None,
+        split=0,
+        train=False,
+        lfs=None,
+        update=False,
+        clear=True,
+        **kwargs
+    ):
+        """Call the LabelerUDF."""
+        if lfs is None:
+            raise ValueError("Please provide a list of labeling functions.")
+
+        if len(lfs) != len(self.candidate_classes):
+            raise ValueError("Please provide LFs for each candidate class.")
+
+        self.lfs = lfs
+        if docs:
```

Comment:
Can we add docstring to mention that doc has higher priority than split?

Reply:
👍 good idea